common:
    lidar_topic:  "/os_cloud_node/points"
    imu_topic:  "/os_cloud_node/imu"
    gnss_topic:  "/gps/fix"
    map_frame:  "map"
    body_frame:  "body"
    timedelay_lidar2imu: 0.0 # Time offset between lidar and IMU calibrated by other algorithms, e.g. LI-Init (can be found in README).

preprocess:
    lidar_type: 3                # 1 for Livox serials LiDAR, 2 for Velodyne LiDAR, 3 for ouster LiDAR, 
    scan_line: 64
    timestamp_unit: 3                 # 0-second, 1-milisecond, 2-microsecond, 3-nanosecond.
    blind: 4
    det_range: 250.0

mapping:
    gravity_align: true                 # true to align the z axis of world frame with the direction of gravity, and the gravity direction should be specified below
    gravity_init: [0, 0, -9.81]         # gravity to be aligned
    eular_init: [0, 0, 0]             # imu init rotation, for fix imu measure
    frontend_type: 0                    # frontend type

    # fastlio
    point_skip_num: 4
    space_down_sample: true
    surf_frame_ds_res: 0.5
    max_iteration: 3
    ikdtree_resolution: 0.5
    lidar_model_search_range: 5
    lidar_meas_cov: 0.001
    cube_side_length: 1000
    runtime_log_enable: 0

    # for backend
    keyframe_add_dist_threshold: 1
    keyframe_add_angle_threshold: 0.2
    recontruct_kdtree: true
    ikdtree_reconstruct_keyframe_num: 10
    ikdtree_reconstruct_downsamp_size: 0.1

    # for add GNSS factor
    lla: [0, 0, 0]
    numsv: 20
    rtk_age: 30
    gpsCovThreshold: [0.05, 0.05, 0.1, 0.05, 0.05, 0.05]
    pose_cov_threshold: 0.
    gnss_weight: [1, 1, 0.1]
    gnssValidInterval: 0.1
    useGpsElevation: true

    # for loop closure
    loop_closure_enable_flag: true
    loop_keyframe_num_thld: 50
    loop_closure_search_radius: 3
    loop_closure_keyframe_interval: 30
    keyframe_search_num: 20
    loop_closure_fitness_score_thld: 0.2
    icp_downsamp_size: 0.1
    manually_loop_vaild_period: [0, 1]
    odom_loop_vaild_period: []
    scancontext_loop_vaild_period: []

    acc_cov: 0.1
    gyr_cov: 0.1
    b_acc_cov: 0.0001
    b_gyr_cov: 0.0001
    extrinsic_est_en:  false      # true: enable the online estimation of IMU-LiDAR extrinsic
    # lidar -> imu = lidar->base * base->imu
    extrinsic_T: [ 0.0, 0.0, 0.0 ]
    extrinsic_R: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1 ]
    # lidar -> gnss
    extrinsic_gnss_T: [ 0, 0, 0 ]
    extrinsic_gnss_R: [ 1, 0, 0, 
                        0, 1, 0, 
                        0, 0, 1 ]

scan_context:
    lidar_height: 2
    sc_dist_thres: 0.13

publish:
    path_en:  true
    scan_publish_en:  true       # false: close all the point cloud output
    dense_publish_en: true       # false: low down the points number in a global-frame point clouds scan.

official:
    ground_constraint_enable: false
    save_globalmap_en: true
    save_resolution: 0.1
    save_keyframe_en: false
    save_keyframe_descriptor_en: false
